src/java.desktop/macosx/native/libawt_lwawt/awt/AWTWindow.m

```diff
@@ -1158,6 +1159,16 @@ + (AWTWindow *) lastKeyWindow {
                     screenContentRect.size.width,
                     screenContentRect.size.height);
                 nsWindow.contentView.frame = contentFrame;
+                if ([nsWindow.contentView.layer isKindOfClass:[MTLLayer class]]) {
+                    MTLLayer* layer = (MTLLayer*)nsWindow.contentView.layer;
+                    if (IS(newBits, FULL_WINDOW_CONTENT)) {
+                        layer.topInset = 0;
+                        layer.leftInset = 0;
+                    } else {
+                        layer.topInset = (jint)(frame.size.height - screenContentRect.size.height);
+                        layer.leftInset = (jint)(screenContentRect.origin.x - frame.origin.x);
+                    }
+                }
```

Comment:
Can you check that it will work in the "new" tabbed mode on big sur as well? Probably it is possible to swap coordinates during rendering in the layer and get rid of this field?

Reply:
> Can you check that it will work in the "new" tabbed mode on big sur as well?

Could you suggest any working test scenario? My simple test (below) just hangs even with OGL:

import java.awt.FlowLayout;
import java.util.Objects;
import javax.swing.*;
import javax.swing.border.EmptyBorder;

public class TestWindowInsets
  extends JDialog
{

    public TestWindowInsets()
    {
        JComponent contentPane = (JComponent) getContentPane();
        contentPane.setBorder(new EmptyBorder(50, 50, 50, 50));
        JButton b = new JButton("Test");
        b.addActionListener(e -> toggle());
        add(b);
        JButton c = new JButton("Win");
        c.addActionListener(e -> win());
        add(c);

        setLayout(new FlowLayout());
        setSize(800, 600);

        setVisible(true);

    }

    void toggle()
    {
        SwingUtilities.invokeLater(() -> {
            JRootPane rp = getRootPane();
            String name = "apple.awt.fullWindowContent";
            Object value = rp.getClientProperty(name);
            if (Objects.equals(value, "true")) {
                value = "false";
            } else {
                value = "true";
            }
            rp.putClientProperty(name, value);
        });
    }

    void win()
    {
        SwingUtilities.invokeLater(TestWindowInsets::new);
    }

    public static void main(String[] args)
    {
        SwingUtilities.invokeLater(TestWindowInsets::new);
    }
}

Reply:
> Probably it is possible to swap coordinates during rendering in the layer and get rid of this field?

I don't see a way how we can do it. Because of the inverted y coordinate in metal, we need to place the origin of the drawing with the appropriate offset that depends on the title height, even if we invert the y coordinate to match java2d coordinate system.

Reply:
I am just not sure that this is the only place where the insets might be changed.

We need to add a "callback" which will be called every time the insets will be changed. Probably in the MTLLayer.replaceSurfaceData which should be called from the LWWindowsPeer#notifyReshape->replaceSurfaceData(). Currently, we skip that call if insets were changed. and this is why this bug arise.